Src/ILGPU/Runtime/OpenCL/CLAccelerator.cs

```diff
@@ -543,8 +546,31 @@ protected override CLKernel CreateKernel(CLCompiledKernel compiledKernel)
             new CLStream(this, ptr, responsible);
 
         /// <summary cref="Accelerator.Synchronize"/>
-        protected override void SynchronizeInternal() =>
-            DefaultStream.Synchronize();
+        protected unsafe override void SynchronizeInternal()
+        {
+            // This is a bad way to do this, but the other alternative,
+            // maintaining a count of specifically stream objects is
+            // AcceleratorObject wasn't so clean either. There has to
+            // be a better way, I just don't know what it is.
+            int numOfStreams = 0;
+            ForEachChildObject((CLStream obj) => numOfStreams++);
+
+            // All of the following feels like it should be in a fixed context,
+            // but I can't find a place where the compiler will let me
+            // create one, so I assume it's not needed?
+            IntPtr*[] arr = new IntPtr*[numOfStreams];
+            int index = 0;
+            ForEachChildObject((CLStream obj) =>
+            {
+                // Setting the wait list count to 0 and the wait list to null
+                // will wait for all operations enqueued prior to this one
+                CLAPI.BarrierWithWaitList(obj.CommandQueue, 0, null, arr[index]);
+                index++;
+            });
+            // We use all the events we obtained from BarrierWithWaitList to wait for all
+            // queued operations to finish
+            CLAPI.WaitForEvents(arr.Length, arr[0]);
```

Comment:
SUGGESTION: You can do this in one pass and maintain a list outside the `ForEachChildObject` loop.

Regarding the IntPtr array, you can use `stackalloc` to allocate a low-cost buffer.

```suggestion
            var streamEvents = new List<IntPtr>();
            try
            {
                ForEachChildObject<CLStream>(stream =>
                {
                    IntPtr* resultEvent = stackalloc IntPtr[1];
                    CLException.ThrowIfFailed(
                        CLAPI.EnqueueBarrierWithWaitList(
                            stream.CommandQueue,
                            Array.Empty<IntPtr>(),
                            resultEvent));
                    streamEvents.Add(*resultEvent);
                });
                
                if (streamEvents.Count > 0)
                    CLException.ThrowIfFailed(CLAPI.WaitForEvents(streamEvents.ToArray()));
            }
            finally
            {
                foreach (var streamEvent in streamEvents)
                    CLAPI.ReleaseEvent(streamEvent);
            }
```

Reply:
Good idea. What I was struggling with originally was maintaining a list of `IntPtr*`, which C# wouldn't let me do. I don't know why I didn't think of just dereferencing the pointers.

Reply:
One question though. Is there a difference in passing `Array.Empty` or null to the API? The documentation said it should be null, that's why I'm curious.

Reply:
Not much difference. When you use the `fixed` statement on `Array.Empty`, it returns null - so that satisfies the OpenCL API.

Reply:
I see. `Array.Empty` seems good then.